### DISCUSSION TITLE
replication - unclear wiki

### DISCUSSION THREAD

**Discussion by schmitch** [seaweedfs/seaweedfs#2385] [Q&A] [answered]
I'm trying to evaulate seaweedfs for our setup.
currently we mostly use a 3 node setup.

however I've found the wiki confusing regarding the replication feature. we prefer to have the files on all 3 nodes, so that each node contains the same files (basically)

> https://github.com/chrislusf/seaweedfs/wiki/Replication
> SeaweedFS can support replication. The replication is implemented not on file level, but on volume level.

however the volume level is meant the "volume server" and not the "volume" per server?
because if I use replication `2` and a single node goes down I can't make more reads, is this correct?
so I would either need to change the replication of new files or bring the server up again?

what is prefered in that case? using replication `1` and rebalance if one server dies or keep replication to `2` and restart weedfs once a server died with replication `1 ` for new files? downtime is not a problem availability of files is.

--

edit:
ok, if one node is dead when started with `replication=002` it won't even allow me to call:
```
curl "http://localhost:9333/dir/assign?replication=001"
{"error":"No free volumes left for {\"replication\":{\"node\":1},\"ttl\":{\"Count\":0,\"Unit\":0}}"}%
```

is there a way around that?
I guess the only way would be to make the volume replication via `volume.configure.replication` and after that I need to change it back again after the node is up again or am I wrong?

<strike>also what about rebalancing if a node is up again? is this possible? i.e. volume replication was 1 and I want to rebalance it to 2? </strike>
ok I found it, it's possible: https://github.com/chrislusf/seaweedfs/wiki/Replication#fix-replication

**Comment by chrislusf** [comment, marked as answer]
> however the volume level is meant the "volume server" and not the "volume" per server? because if I use replication `2` and a single node goes down I can't make more reads, is this correct? so I would either need to change the replication of new files or bring the server up again?
> 

1. Yes.
2. No. Writes are not allowed on those volumes. But reads will still work.
3. New files will be assigned to other volumes.

**Comment by chrislusf** [comment]
> ok, if one node is dead when started with `replication=002` it won't even allow me to call:
> 
> ```
> curl "http://localhost:9333/dir/assign?replication=001"
> {"error":"No free volumes left for {\"replication\":{\"node\":1},\"ttl\":{\"Count\":0,\"Unit\":0}}"}%
> ```
> 
> is there a way around that?

Fix the replication or use 4 or more volume servers to allow one volume server offline and if you must use replication `002`. Basically, if you need to write with `002`, you need to have at least 3 volume servers online.

**Comment by chrislusf** [comment]
Seems you have many questions around this. Not sure whether I answered all of them. But when you are clear, could you please help to adjust the wiki to make it clearer? It's open to write.

**Reply by schmitch** [reply]
well yeah you mostly cleared them up!
btw. I still have a simple question, currently I also tried the s3 interface. And I try to upload files in chuncks of 5Mb which works, but each chunck upload gets slower and slower.
I created a 10G file via:

```
head -c 10737418240 </dev/urandom >myfile.txt
```
and each upload takes more time. i.e. the first chunck takes like 308ms (warmup) than 80ms than 89ms, than 130ms uvm so it grows linear (chunck 152 already takes like 4s)
is this normal? 

I started it with: `./weed server -s3 -ip=localhost -master.port=9333 -dir=./x1 -volume.port=8080 5 -master.defaultReplication=000 -dataCenter=dc1 -rack=rack1 -volume.max=0 -index=leveldb3 -filer.maxMB=5` and I'm on a mac (nvme drive)
it also clogs the cpu (it might be a macos bug tough)

I wanted to test s3 aswell since it's easier to do backups via the `filer` however in theory we would not need it.
it also took way more storage for each chunck. i.e. I'm nearly done and the upload took like 80gig

**Reply by chrislusf** [reply]
Seems you are using Fuse mount to upload the big file, which needs to copy the file to local first and then upload. 

You can try to use "weed filer.copy", or s3 multi part upload.

**Reply by schmitch** [reply]
I actually used s3 multi part upload, I've written a small c# programm which created the multi part upload https://gist.github.com/schmitch/798525b13ca14e9970288d1392412575

**Reply by chrislusf** [reply]
Maybe some code bug. Try the "aws s3" command to upload.

**Reply by schmitch** [reply]
looks like I made a stupid mistake, still thanks for your time.
seaweedfs looks working really really well.

**Reply by chrislusf** [reply]
thanks for the update! Curious to know what is the bug. I eye-balled the code and could not find the bug.

**Reply by schmitch** [reply]
RangeLength was RangeEnd. this explained the odd "gets bigger" and slower over time.

btw. I still have a bug, I uploaded 10G but the uploaded file is only 5G big, but this happens with aws cli aswell (the file is just a little bit bigger 8G instead of 5G)
is this a limitation because it uses too many chuncks? i.e. the default chunck size of aws cli is 8MB i guess and I used 5MB which would explain it: https://docs.aws.amazon.com/cli/latest/topic/s3-config.html#multipart-chunksize

```
➜  weedfs /Users/schmitch/Applications/aws/aws-cli/aws --endpoint-url http://localhost:8333 s3 mb s3://hase
make_bucket: hase
➜  weedfs ls -l
-rw-r--r--   1 schmitch  staff  10737418240 15 Okt 11:29 myfile.text
➜  weedfs /Users/schmitch/Applications/aws/aws-cli/aws --endpoint-url http://localhost:8333 s3 cp myfile.txt s3://hase/hase.txt
upload: ./myfile.txt to s3://hase/hase.txt
➜  weedfs /Users/schmitch/Applications/aws/aws-cli/aws --endpoint-url http://localhost:8333 s3 ls s3://hase
2021-10-18 22:18:35 8388608000 hase.txt
➜  weedfs /Users/schmitch/Applications/aws/aws-cli/aws --endpoint-url http://localhost:8333 s3 ls s3://hase
2021-10-18 22:18:35 8388608000 hase.txt
```

**Reply by chrislusf** [reply]
You are not hitting the limit. There is a limit of 10000 parts, not 1000. So the upload file size is 50GB if using 5MB chunk.

**Reply by schmitch** [reply]
hm as said what's strange is that even the aws-cli returns that result, just with 8gb, however mounting the filer and copying the same file works correctly.

**Reply by chrislusf** [reply]
ok. I can reproduce this locally. Will keep you updated.

**Reply by chrislusf** [reply]
which SeaweedFS version are you using btw?

**Reply by schmitch** [reply]
I tried 2.71 but I upgraded to 2.73 to double check (and recreated the volumes), but it happens there aswell. on macos

**Reply by chrislusf** [reply]
found the problem. Fixed it in https://github.com/chrislusf/seaweedfs/commit/309f46e3af0c4eaba3f5a4f632aad0dd87be1f93
